web/src/user_group_edit_members.ts

```diff
@@ -258,17 +352,56 @@ function add_new_members({pill_user_ids}: {pill_user_ids: number[]}): void {
         );
     }
 
-    if (user_id_set.size === 0) {
+    const deactivated_groups = new Set<number>();
+    const already_added_subgroups = new Set<number>();
+
+    const existing_subgroup_ids = new Set(group.direct_subgroup_ids);
+    const subgroup_ids_to_add = pill_group_ids.filter((group_id) => {
+        const subgroup = user_groups.get_user_group_from_id(group_id);
+        if (subgroup.deactivated) {
+            deactivated_groups.add(group_id);
+            return false;
+        }
+
+        if (existing_subgroup_ids.has(group_id)) {
+            already_added_subgroups.add(group_id);
+            return false;
+        }
+
+        return true;
+    });
+
+    let ignored_deactivated_groups: UserGroup[] = [];
+    let ignored_already_added_subgroups: UserGroup[] = [];
+    if (deactivated_groups.size > 0) {
+        const ignored_deactivated_group_ids = [...deactivated_groups];
+        ignored_deactivated_groups = ignored_deactivated_group_ids.map((group_id) =>
+            user_groups.get_user_group_from_id(group_id),
+        );
+    }
+    if (already_added_subgroups.size > 0) {
+        const ignored_already_added_subgroup_ids = [...already_added_subgroups];
+        ignored_already_added_subgroups = ignored_already_added_subgroup_ids.map((group_id) =>
+            user_groups.get_user_group_from_id(group_id),
+        );
```

Comment:
This can probably be deduplicated. Or maybe we can split some code to different functions for readability. Will look into this.

Reply:
Yeah this could be nicer but it's basically fine from a readability standpoint.